upapap

```diff
@@ -1,49 +1,146 @@
 #!/usr/bin/env python3
 # encoding: utf-8 (as per PEP 263)
 
+import abc
 import sys
 
 import requests
 from bs4 import BeautifulSoup
 
-urls = [
+__all__ = ["URLS", "Callback", "Markdown", "HTML", "parse_lolol", "feed_handler"]
+
+URLS = [
     'https://www.fim.uni-passau.de/studium/modulkataloge/',
     'https://www.fim.uni-passau.de/index.php?id=17010',
     'https://www.fim.uni-passau.de/ueber-die-fakultaet/ausschuesse/pruefungsausschuss/',
     'https://www.fim.uni-passau.de/studium/pruefungsordnungen/',
 ]
 
-def parse_lolol(soup):
+
+class Callback(abc.ABC):
+    @abc.abstractmethod
+    def handle_page_title(self, url, pagetitle):
+        pass
+
+    @abc.abstractmethod
+    def handle_section_title(self, heading):
+        pass
+
+    @abc.abstractmethod
+    def handle_entry(self, href, text):
+        pass
+
+    @abc.abstractmethod
+    def handle_end(self):
+        pass
+
+
+class Markdown(Callback):
+    def __init__(self):
+        self.response = ""
+        self.last_heading = ""
+
+    def handle_page_title(self, url, pagetitle):
+        if self.response:
+            self.response += '\n'
+        self.response += '# %s\n' % pagetitle
+        self.response += '%s\n' % url
+
+    def handle_section_title(self, heading):
+        if self.last_heading != heading:
+            self.response += '\n## %s\n' % heading
+        self.last_heading = heading
+
+    def handle_entry(self, href, text):
+        if href:
+            self.response += '* %s\n  %s\n' % (text, href)
```

Comment:
The two trailing space characters are intentional ([Markdown syntax to force a newline](https://daringfireball.net/projects/markdown/syntax#p)) and should be preserved.